scalafix-tests/unit/src/test/scala/scalafix/tests/core/util/TokenListTest.scala

```diff
@@ -6,21 +6,90 @@ import scalafix.util.TokenList
 
 class TokenListTest extends FunSuite {
 
-  val tokens = "".tokenize.get // this contains two tokens: beginningOfFile and endOfFile
-  val firstToken = tokens.head
-  val lastToken = tokens.last
-  val tokenList = new TokenList(tokens)
+  val emptyFileTokens = "".tokenize.get // this contains two tokens: BOF and EOF
+  val nonEmptyFileTokens =
+    """package foo
+      |
+      |object Bar {
+      | val baz   =   10
+      |}
+    """.stripMargin.tokenize.get
 
-  test("Prev returns the firstToken when is given the firstToken") {
-    assert(tokenList.prev(firstToken) == firstToken)
-    assert(tokenList.prev(tokens.last) == firstToken)
+  test("prev returns the preceding token") {
+    val tokenList = TokenList(emptyFileTokens)
+    assert(tokenList.prev(emptyFileTokens.last) == emptyFileTokens.head)
   }
 
-  test("Slice gives an empty list with same token as inputs") {
-    assert(tokenList.slice(firstToken, firstToken) == Seq())
+  test("prev returns self if there is no preceding token") {
+    val tokenList = TokenList(emptyFileTokens)
+    assert(tokenList.prev(emptyFileTokens.head) == emptyFileTokens.head)
   }
 
-  test("Next breaks gives the lastToken if it is given lastToken") {
-    assert(tokenList.next(lastToken) == lastToken)
+  test("next returns the following token") {
+    val tokenList = TokenList(emptyFileTokens)
+    assert(tokenList.next(emptyFileTokens.head) == emptyFileTokens.last)
+  }
+
+  test("next returns self if there is no following token") {
+    val tokenList = TokenList(emptyFileTokens)
+    assert(tokenList.next(emptyFileTokens.last) == emptyFileTokens.last)
+  }
+
+  test("slice returns an empty seq if there is no token in between") {
+    val tokenList = TokenList(emptyFileTokens)
+    assert(tokenList.slice(emptyFileTokens.head, emptyFileTokens.head) == Seq())
+  }
+
```

Comment:
it's kinda off topic but I think it's worth mentioning: when I was working on the previous implementation I found 2 things in the API that didn't behave as I expected (http://wiki.c2.com/?PrincipleOfLeastAstonishment):

- `prev`/`next` returns the input token if it is the head/last of the tokenList. Would `Option[Token]` make more sense?
- `slice(from, to)` doesn't behave as in Scala's std collection given that `from` is not inclusive. For instance, while removing tokens what could take a single step (`ctx.removeTokens(tokenList.slice(tk, last)`) actually requires 2 (`ctx.removeToken(tk) + ctx.removeTokens(tokenList.slice(tk, last)`) and is really easy to introduce bugs if users don't dive into the implementation of `slice`

Reply:
These are great points, I agree both of these can be improved. Can you open tickets? I am open to fix `slice` immediately (we can even implement is using slice on views from the stdlib) while prev/next will have to wait until v0.6

Reply:
Awesome! Here are the issues I created:
https://github.com/scalacenter/scalafix/issues/516
https://github.com/scalacenter/scalafix/issues/517